src/main/java/com/rackspace/ceres/app/services/SuggestApiService.java

```diff
@@ -0,0 +1,75 @@
+package com.rackspace.ceres.app.services;
+
+import java.util.List;
+import javax.xml.crypto.Data;
+import jnr.ffi.annotations.Meta;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.cassandra.core.cql.ReactiveCqlTemplate;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+@Service
+public class SuggestApiService {
+
+  private final MetadataService metadataService;
+
+  @Autowired
+  public SuggestApiService(MetadataService metadataService) {
+    this.metadataService = metadataService;
+  }
+
+  /**
+   * Method used to return the list of tag keys associated with particular tenant with limit set to
+   * max
+   *
+   * @param tenant
+   * @param tagK
+   * @param max
+   * @return
+   */
+  public Mono<List<String>> suggestTagKeys(String tenant, String tagK, int max) {
+    return metadataService.getMetricNames(tenant).flatMapMany(Flux::fromIterable)
+        .flatMap(metric -> metadataService.getTagKeys(tenant, metric).flatMapMany(Flux::fromIterable)
+        .filter(tagKey -> !StringUtils.hasText(tagK) || tagKey.startsWith(tagK)))
+        .distinct()
+        .take(max)
+        .collectList();
+  }
+
+  /**
+   * Method used to return the list of metric names that starts with metric for a tenant and limits
+   * the result to max.
+   *
+   * @param tenant
+   * @param metric
+   * @param max
+   * @return
+   */
+  public Mono<List<String>> suggestMetricNames(String tenant, String metric, int max) {
+    return metadataService.getMetricNames(tenant).flatMapMany(Flux::fromIterable)
+        .filter(metricName -> !StringUtils.hasText(metric) || metricName.startsWith(metric))
```

Comment:
I am trying to think how many metricNames we should expect to be on a particular tenant. I guess it shouldn't be more than a few hundred results. So it should be fine to do in code, but this is the type of thing that could also be done in a database query.

Reply:
Actually cassandra does not have support for like query as it is in SQL.

Reply:
We should double check we aren't storing these values in Redis and also do a spike on possibly adding the ability to do this sort of query against Cassandra itself.

Reply:
https://docs.datastax.com/en/dse/6.0/cql/cql/cql_using/search_index/queryTerms.html seems to indicate that its possible but we need to build out a secondary index with Solr to actually be able to do these style of string comparisons. In 10 minutes I haven't been able to find a native way to query for a regex against a primary key in cassandra which seems really weird to me. That is part of the clever sort of naming schemes that Key/Value stores rely upon to be able to speed up data queries in single requests.